codex/blockexchange/engine/engine.nim

```diff
@@ -116,39 +116,66 @@ proc stop*(b: BlockExcEngine) {.async.} =
 
   trace "NetworkStore stopped"
 
+proc sendWantHave(b: BlockExcEngine, cid: Cid, selectedPeer: BlockExcPeerCtx, peers: seq[BlockExcPeerCtx]): Future[void] {.async.} =
+  trace "Sending wantHave request to peers", cid
+  for p in peers:
+    if p != selectedPeer:
+      if cid notin p.peerHave:
+        trace " wantHave > ", peer = p.id
+        await b.network.request.sendWantList(
+          p.id,
+          @[cid],
+          wantType = WantType.WantHave) # we only want to know if the peer has the block
+
+proc sendWantHaveSingle(b: BlockExcEngine, cid: Cid, peer: PeerId): Future[void] {.async.} =
+  trace "Sending wantHave request to single peer", cid, peer
+  await b.network.request.sendWantList(
+    peer,
+    @[cid],
+    wantType = WantType.WantHave) # we only want to know if the peer has the block
+
+proc sendWantBlock(b: BlockExcEngine, cid: Cid, blockPeer: BlockExcPeerCtx): Future[void] {.async.} =
+  trace "Sending wantBlock request to", peer = blockPeer.id, cid
+  await b.network.request.sendWantList(
+    blockPeer.id,
+    @[cid],
+    wantType = WantType.WantBlock) # we want this remote to send us a block
+
+proc findCheapestPeerForBlock(b: BlockExcEngine, cheapestPeers: seq[BlockExcPeerCtx]): ?BlockExcPeerCtx =
+  if cheapestPeers.len <= 0:
+    trace "No cheapest peers, selecting first in list"
+    let peers = toSeq(b.peers) # Get any peer
+    if peers.len <= 0:
+      return none(BlockExcPeerCtx)
+    return some(peers[0])
+  return some(cheapestPeers[0]) # get cheapest
+
 proc requestBlock*(
   b: BlockExcEngine,
   cid: Cid,
   timeout = DefaultBlockTimeout): Future[bt.Block] {.async.} =
-  ## Request a block from remotes
-  ##
-
-  trace "Requesting block", cid, peers = b.peers.len
+  trace "Begin block request", cid, peers = b.peers.len
 
   if b.pendingBlocks.isInFlight(cid):
     trace "Request handle already pending", cid
     return await b.pendingBlocks.getWantHandle(cid, timeout)
 
-  let
-    blk = b.pendingBlocks.getWantHandle(cid, timeout)
-
-  var
-    peers = b.peers.selectCheapest(cid)
+  let blk = b.pendingBlocks.getWantHandle(cid, timeout)
 
+  trace "Selecting peers who have", cid
+  var peers = b.peers.selectCheapest(cid)
   if peers.len <= 0:
-    trace "No cheapest peers, selecting first in list", cid
-    peers = toSeq(b.peers) # Get any peer
-    if peers.len <= 0:
-      trace "No peers to request blocks from", cid
+    trace "We don't know of anyone who has this. Start discovery immediately."
+    b.discovery.queueFindBlocksReq(@[cid])
+
+  without blockPeer =? b.findCheapestPeerForBlock(peers):
```

Comment:
Why spawn discovery twice? There is no reason to do it twice I believe.

Reply:
OK, I reran the test `FullyConnectedDownloadTest(3,1)` (cs-codex-dist-tests repository, branch feature/fully-connected-download-tests) with and without these lines:

```
if peers.len <= 0:
  trace "We don't know of anyone who has this. Start discovery immediately."
  b.discovery.queueFindBlocksReq(@[cid])
```

With lines:
`Performed 12 downloads of 1MB in 13.8697523 seconds, for an average of 00:00:01.1558127 seconds per MB.` (Test passed)

Without lines:
`Performed 12 downloads of 1MB in 427.2357426 seconds, for an average of 00:00:35.6029786 seconds per MB.` (Test failed)

I agree the code looks like it's just duplicate. But somehow, there's a case where this call makes the whole thing many times faster. (There's a performance-threshold assert in the tests that cause the tests to fail when download is this slow.)

I don't understand exactly who it causes this difference. But the fact that it's close to 30 seconds, and the discovery-loop sleeps for 30 seconds is very... suspicious... Also I propose I don't put too much time into this mystery, since the design for the future block exchange engine looks different anyway. What do you think?

Reply:
I moved the `if ... queueFindBlocksReq` back down and reran the tests and indeed that still works fine.